File: src/mcp/tools/column.ts
// src/mcp/tools/column.ts
// MCP tools: create_column, update_column, delete_column, reorder_columns

import { z } from 'zod'
import { assertProjectAccess, getAuthedUserId } from '../auth'
import { socketEmitWithAck } from '../socket-manager'
import {
  McpError,
  ackCodeToMcpCode,
  makeMcpError,
  makeMcpSuccess,
  toMcpErrorResponse,
} from '../errors'
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js'
import { dataTypeSchema } from '@/data/schema'
import { getTableProjectId } from '@/data/resolve-project'
import { findDiagramTableById } from '@/data/diagram-table'
import { findColumnById } from '@/data/column'

/**
 * Get the whiteboard ID for a table (needed to get the socket connection namespace).
 */
async function getWhiteboardIdForTable(tableId: string): Promise<string> {
  const table = await findDiagramTableById(tableId)
  if (!table) throw new McpError('NOT_FOUND', `Table ${tableId} not found.`)
  return table.whiteboardId
}


export function registerColumnTools(server: McpServer): void {
  // -------------------------------------------------------------------------
  // create_column
  // -------------------------------------------------------------------------
  server.tool(
    'create_column',
    'Create a new column in a table.',
    {
      tableId: z.string().uuid().describe('The table UUID'),
      name: z.string().min(1).max(255).describe('Column name'),
      dataType: dataTypeSchema.describe('Column data type'),
      isPrimaryKey: z.boolean().optional().describe('Is primary key'),
      isForeignKey: z.boolean().optional().describe('Is foreign key'),
      isUnique: z.boolean().optional().describe('Is unique'),
      isNullable: z.boolean().optional().describe('Is nullable'),
      description: z.string().optional().describe('Optional description'),
      order: z
        .number()
        .int()
        .min(0)
        .optional()
        .describe('Column order position'),
    },
    async ({
      tableId,
      name,
      dataType,
      isPrimaryKey,
      isForeignKey,
      isUnique,
      isNullable,
      description,
      order,
    }) => {
      try {
        const userId = getAuthedUserId()
        const projectId = await getTableProjectId(tableId)
        if (!projectId)
          throw new McpError('NOT_FOUND', `Table ${tableId} not found.`)
        await assertProjectAccess(userId, projectId)

        const whiteboardId = await getWhiteboardIdForTable(tableId)
        const payload: Record<string, unknown> = { tableId, name, dataType }
        if (isPrimaryKey !== undefined) payload.isPrimaryKey = isPrimaryKey
        if (isForeignKey !== undefined) payload.isForeignKey = isForeignKey
        if (isUnique !== undefined) payload.isUnique = isUnique
        if (isNullable !== undefined) payload.isNullable = isNullable
        if (description !== undefined) payload.description = description
        if (order !== undefined) payload.order = order

        const ack = await socketEmitWithAck(
          whiteboardId,
          'column:create',
          payload,
        )
        if (ack.ok === false) {
          return makeMcpError(
            ackCodeToMcpCode(ack.code as string),
            (ack.message as string | undefined) ?? 'Server rejected column creation.',
          )
        }
        return makeMcpSuccess(ack.entity)
      } catch (err) {
        return toMcpErrorResponse(err)
      }
    },
  )

  // -------------------------------------------------------------------------
  // update_column
  // -------------------------------------------------------------------------
  server.tool(
    'update_column',
    "Update a column's properties.",
    {
      columnId: z.string().uuid().describe('The column UUID'),
      name: z.string().min(1).max(255).optional().describe('New name'),
      dataType: dataTypeSchema.optional().describe('New data type'),
      isPrimaryKey: z.boolean().optional().describe('Is primary key'),
      isForeignKey: z.boolean().optional().describe('Is foreign key'),
      isUnique: z.boolean().optional().describe('Is unique'),
      isNullable: z.boolean().optional().describe('Is nullable'),
      description: z.string().optional().describe('New description'),
    },
    async ({
      columnId,
      name,
      dataType,
      isPrimaryKey,
      isForeignKey,
      isUnique,
      isNullable,
      description,
    }) => {
      try {
        const userId = getAuthedUserId()
        const column = await findColumnById(columnId)
        if (!column)
          throw new McpError('NOT_FOUND', `Column ${columnId} not found.`)

        const projectId = await getTableProjectId(column.tableId)
        if (!projectId)
          throw new McpError('NOT_FOUND', `Column ${columnId} not found.`)
        await assertProjectAccess(userId, projectId)

        const whiteboardId = await getWhiteboardIdForTable(column.tableId)
        const payload: Record<string, unknown> = { columnId }
        if (name !== undefined) payload.name = name
        if (dataType !== undefined) payload.dataType = dataType
        if (isPrimaryKey !== undefined) payload.isPrimaryKey = isPrimaryKey
        if (isForeignKey !== undefined) payload.isForeignKey = isForeignKey
        if (isUnique !== undefined) payload.isUnique = isUnique
        if (isNullable !== undefined) payload.isNullable = isNullable
        if (description !== undefined) payload.description = description

        const ack = await socketEmitWithAck(
          whiteboardId,
          'column:update',
          payload,
        )
        if (ack.ok === false) {
          return makeMcpError(
            ackCodeToMcpCode(ack.code as string),
            (ack.message as string | undefined) ?? 'Server rejected column update.',
          )
        }
        return makeMcpSuccess(ack.entity)
      } catch (err) {
        return toMcpErrorResponse(err)
      }
    },
  )

  // -------------------------------------------------------------------------
  // delete_column
  // -------------------------------------------------------------------------
  server.tool(
    'delete_column',
    'Delete a column. Returns the deleted column ID and cascade relationship count.',
    {
      columnId: z.string().uuid().describe('The column UUID'),
    },
    async ({ columnId }) => {
      try {
        const userId = getAuthedUserId()
        const column = await findColumnById(columnId)
        if (!column)
          throw new McpError('NOT_FOUND', `Column ${columnId} not found.`)

        const projectId = await getTableProjectId(column.tableId)
        if (!projectId)
          throw new McpError('NOT_FOUND', `Column ${columnId} not found.`)
        await assertProjectAccess(userId, projectId)

        const whiteboardId = await getWhiteboardIdForTable(column.tableId)
        const ack = await socketEmitWithAck(whiteboardId, 'column:delete', {
          columnId,
        })
        if (ack.ok === false) {
          return makeMcpError(
            ackCodeToMcpCode(ack.code as string),
            (ack.message as string | undefined) ?? 'Server rejected column deletion.',
          )
        }

        const cascade = ack.cascade as { relationships?: number } | undefined
        const warnings: Array<string> = []
        if (cascade?.relationships && cascade.relationships > 0) {
          warnings.push(`${cascade.relationships} relationship(s) deleted`)
        }

        return makeMcpSuccess({
          id: columnId,
          cascade,
          warning:
            warnings.length > 0
              ? `Cascade deleted: ${warnings.join(', ')}.`
              : undefined,
        })
      } catch (err) {
        return toMcpErrorResponse(err)
      }
    },
  )

  // -------------------------------------------------------------------------
  // reorder_columns
  // -------------------------------------------------------------------------
  server.tool(
    'reorder_columns',
    'Reorder columns within a table by providing the desired column ID order.',
    {
      tableId: z.string().uuid().describe('The table UUID'),
      orderedColumnIds: z
        .array(z.string().uuid())
        .min(1)
        .max(500)
        .describe('Column IDs in the desired order'),
    },
    async ({ tableId, orderedColumnIds }) => {
      try {
        const userId = getAuthedUserId()
        const projectId = await getTableProjectId(tableId)
        if (!projectId)
          throw new McpError('NOT_FOUND', `Table ${tableId} not found.`)
        await assertProjectAccess(userId, projectId)

        const whiteboardId = await getWhiteboardIdForTable(tableId)
        const ack = await socketEmitWithAck(whiteboardId, 'column:reorder', {
          tableId,
          orderedColumnIds,
        })
        if (ack.ok === false) {
          return makeMcpError(
            ackCodeToMcpCode(ack.code as string),
            (ack.message as string | undefined) ?? 'Server rejected column reorder.',
          )
        }
        return makeMcpSuccess(ack.entity ?? { tableId, orderedColumnIds })
      } catch (err) {
        return toMcpErrorResponse(err)
      }
    },
  )
}

File: internal/tools/column.go
package tools

import (
	"context"
	"fmt"
	"strings"

	"github.com/modelcontextprotocol/go-sdk/mcp"

	"github.com/LizardLiang/liz-whiteboard-mcp/internal/auth"
	"github.com/LizardLiang/liz-whiteboard-mcp/internal/data"
	mcperr "github.com/LizardLiang/liz-whiteboard-mcp/internal/errors"
	"github.com/LizardLiang/liz-whiteboard-mcp/internal/schema"
	"github.com/LizardLiang/liz-whiteboard-mcp/internal/socket"
)

type createColumnInput struct {
	TableID      string  `json:"tableId" jsonschema:"The table UUID"`
	Name         string  `json:"name" jsonschema:"Column name"`
	DataType     string  `json:"dataType" jsonschema:"Column data type"`
	IsPrimaryKey *bool   `json:"isPrimaryKey,omitempty" jsonschema:"Is primary key"`
	IsForeignKey *bool   `json:"isForeignKey,omitempty" jsonschema:"Is foreign key"`
	IsUnique     *bool   `json:"isUnique,omitempty" jsonschema:"Is unique"`
	IsNullable   *bool   `json:"isNullable,omitempty" jsonschema:"Is nullable"`
	Description  *string `json:"description,omitempty" jsonschema:"Optional description"`
	Order        *int    `json:"order,omitempty" jsonschema:"Column order position"`
}

type updateColumnInput struct {
	ColumnID     string  `json:"columnId" jsonschema:"The column UUID"`
	Name         *string `json:"name,omitempty" jsonschema:"New name"`
	DataType     *string `json:"dataType,omitempty" jsonschema:"New data type"`
	IsPrimaryKey *bool   `json:"isPrimaryKey,omitempty" jsonschema:"Is primary key"`
	IsForeignKey *bool   `json:"isForeignKey,omitempty" jsonschema:"Is foreign key"`
	IsUnique     *bool   `json:"isUnique,omitempty" jsonschema:"Is unique"`
	IsNullable   *bool   `json:"isNullable,omitempty" jsonschema:"Is nullable"`
	Description  *string `json:"description,omitempty" jsonschema:"New description"`
}

type columnIDInput struct {
	ColumnID string `json:"columnId" jsonschema:"The column UUID"`
}

type reorderColumnsInput struct {
	TableID          string   `json:"tableId" jsonschema:"The table UUID"`
	OrderedColumnIDs []string `json:"orderedColumnIds" jsonschema:"Column IDs in the desired order"`
}

// getWhiteboardIDForTable resolves the whiteboard ID for a table, returning a
// NOT_FOUND McpError if the table does not exist.
func getWhiteboardIDForTable(ctx context.Context, tableID string) (string, error) {
	table, err := data.FindDiagramTableByID(ctx, tableID)
	if err != nil {
		return "", err
	}
	if table == nil {
		return "", mcperr.New(mcperr.NotFound, fmt.Sprintf("Table %s not found.", tableID))
	}
	return table.WhiteboardID, nil
}

// RegisterColumnTools registers create_column, update_column, delete_column,
// reorder_columns.
func RegisterColumnTools(s *mcp.Server) {
	registerCreateColumn(s)
	registerUpdateColumn(s)
	registerDeleteColumn(s)
	registerReorderColumns(s)
}

func registerCreateColumn(s *mcp.Server) {
	mcp.AddTool(s, &mcp.Tool{
		Name:        "create_column",
		Description: "Create a new column in a table.",
	}, func(ctx context.Context, _ *mcp.CallToolRequest, in createColumnInput) (*mcp.CallToolResult, any, error) {
		if e := validateUUID("tableId", in.TableID); e != nil {
			return fail(e)
		}
		if e := checkLen("name", in.Name, 1, 255); e != nil {
			return fail(e)
		}
		if !schema.IsValidDataType(in.DataType) {
			return validationError("Invalid enum value.", "dataType")
		}
		// W6: order z.number().int().min(0)
		if in.Order != nil {
			if e := checkMinOrder("order", *in.Order); e != nil {
				return fail(e)
			}
		}

		userID := auth.UserID(ctx)
		projectID, err := data.GetTableProjectID(ctx, in.TableID)
		if err != nil {
			return fail(err)
		}
		if projectID == "" {
			return mcpError(mcperr.NotFound, fmt.Sprintf("Table %s not found.", in.TableID))
		}
		if err := auth.AssertProjectAccess(ctx, userID, projectID); err != nil {
			return fail(err)
		}

		whiteboardID, err := getWhiteboardIDForTable(ctx, in.TableID)
		if err != nil {
			return fail(err)
		}

		payload := map[string]any{"tableId": in.TableID, "name": in.Name, "dataType": in.DataType}
		if in.IsPrimaryKey != nil {
			payload["isPrimaryKey"] = *in.IsPrimaryKey
		}
		if in.IsForeignKey != nil {
			payload["isForeignKey"] = *in.IsForeignKey
		}
		if in.IsUnique != nil {
			payload["isUnique"] = *in.IsUnique
		}
		if in.IsNullable != nil {
			payload["isNullable"] = *in.IsNullable
		}
		if in.Description != nil {
			payload["description"] = *in.Description
		}
		if in.Order != nil {
			payload["order"] = *in.Order
		}

		ack, err := socket.SocketEmitWithAck(ctx, whiteboardID, userID, "column:create", payload)
		if err != nil {
			return fail(err)
		}
		if !ack.OK() {
			return mcpError(mcperr.AckCodeToMcpCode(ack.Code()), msgOr(ack.Message(), "Server rejected column creation."))
		}
		return success(ack.Entity())
	})
}

func registerUpdateColumn(s *mcp.Server) {
	mcp.AddTool(s, &mcp.Tool{
		Name:        "update_column",
		Description: "Update a column's properties.",
	}, func(ctx context.Context, _ *mcp.CallToolRequest, in updateColumnInput) (*mcp.CallToolResult, any, error) {
		if e := validateUUID("columnId", in.ColumnID); e != nil {
			return fail(e)
		}
		// W5: name z.string().min(1).max(255).optional()
		if in.Name != nil {
			if e := checkLen("name", *in.Name, 1, 255); e != nil {
				return fail(e)
			}
		}
		if in.DataType != nil && !schema.IsValidDataType(*in.DataType) {
			return validationError("Invalid enum value.", "dataType")
		}

		userID := auth.UserID(ctx)
		column, err := data.FindColumnByID(ctx, in.ColumnID)
		if err != nil {
			return fail(err)
		}
		if column == nil {
			return mcpError(mcperr.NotFound, fmt.Sprintf("Column %s not found.", in.ColumnID))
		}
		projectID, err := data.GetTableProjectID(ctx, column.TableID)
		if err != nil {
			return fail(err)
		}
		if projectID == "" {
			return mcpError(mcperr.NotFound, fmt.Sprintf("Column %s not found.", in.ColumnID))
		}
		if err := auth.AssertProjectAccess(ctx, userID, projectID); err != nil {
			return fail(err)
		}

		whiteboardID, err := getWhiteboardIDForTable(ctx, column.TableID)
		if err != nil {
			return fail(err)
		}

		payload := map[string]any{"columnId": in.ColumnID}
		if in.Name != nil {
			payload["name"] = *in.Name
		}
		if in.DataType != nil {
			payload["dataType"] = *in.DataType
		}
		if in.IsPrimaryKey != nil {
			payload["isPrimaryKey"] = *in.IsPrimaryKey
		}
		if in.IsForeignKey != nil {
			payload["isForeignKey"] = *in.IsForeignKey
		}
		if in.IsUnique != nil {
			payload["isUnique"] = *in.IsUnique
		}
		if in.IsNullable != nil {
			payload["isNullable"] = *in.IsNullable
		}
		if in.Description != nil {
			payload["description"] = *in.Description
		}

		ack, err := socket.SocketEmitWithAck(ctx, whiteboardID, userID, "column:update", payload)
		if err != nil {
			return fail(err)
		}
		if !ack.OK() {
			return mcpError(mcperr.AckCodeToMcpCode(ack.Code()), msgOr(ack.Message(), "Server rejected column update."))
		}
		return success(ack.Entity())
	})
}

func registerDeleteColumn(s *mcp.Server) {
	mcp.AddTool(s, &mcp.Tool{
		Name:        "delete_column",
		Description: "Delete a column. Returns the deleted column ID and cascade relationship count.",
	}, func(ctx context.Context, _ *mcp.CallToolRequest, in columnIDInput) (*mcp.CallToolResult, any, error) {
		if e := validateUUID("columnId", in.ColumnID); e != nil {
			return fail(e)
		}
		userID := auth.UserID(ctx)
		column, err := data.FindColumnByID(ctx, in.ColumnID)
		if err != nil {
			return fail(err)
		}
		if column == nil {
			return mcpError(mcperr.NotFound, fmt.Sprintf("Column %s not found.", in.ColumnID))
		}
		projectID, err := data.GetTableProjectID(ctx, column.TableID)
		if err != nil {
			return fail(err)
		}
		if projectID == "" {
			return mcpError(mcperr.NotFound, fmt.Sprintf("Column %s not found.", in.ColumnID))
		}
		if err := auth.AssertProjectAccess(ctx, userID, projectID); err != nil {
			return fail(err)
		}

		whiteboardID, err := getWhiteboardIDForTable(ctx, column.TableID)
		if err != nil {
			return fail(err)
		}

		ack, err := socket.SocketEmitWithAck(ctx, whiteboardID, userID, "column:delete", map[string]any{"columnId": in.ColumnID})
		if err != nil {
			return fail(err)
		}
		if !ack.OK() {
			return mcpError(mcperr.AckCodeToMcpCode(ack.Code()), msgOr(ack.Message(), "Server rejected column deletion."))
		}

		cascadeVal := ack["cascade"]
		var warnings []string
		if n := cascadeCount(cascadeVal, "relationships"); n > 0 {
			warnings = append(warnings, fmt.Sprintf("%d relationship(s) deleted", n))
		}
		out := map[string]any{"id": in.ColumnID}
		if cascadeVal != nil {
			out["cascade"] = cascadeVal
		}
		if len(warnings) > 0 {
			out["warning"] = "Cascade deleted: " + strings.Join(warnings, ", ") + "."
		}
		return success(out)
	})
}

func registerReorderColumns(s *mcp.Server) {
	mcp.AddTool(s, &mcp.Tool{
		Name:        "reorder_columns",
		Description: "Reorder columns within a table by providing the desired column ID order.",
	}, func(ctx context.Context, _ *mcp.CallToolRequest, in reorderColumnsInput) (*mcp.CallToolResult, any, error) {
		if e := validateUUID("tableId", in.TableID); e != nil {
			return fail(e)
		}
		if len(in.OrderedColumnIDs) < 1 || len(in.OrderedColumnIDs) > 500 {
			return validationError("orderedColumnIds must contain between 1 and 500 entries.", "orderedColumnIds")
		}
		for _, id := range in.OrderedColumnIDs {
			if e := validateUUID("orderedColumnIds", id); e != nil {
				return fail(e)
			}
		}

		userID := auth.UserID(ctx)
		projectID, err := data.GetTableProjectID(ctx, in.TableID)
		if err != nil {
			return fail(err)
		}
		if projectID == "" {
			return mcpError(mcperr.NotFound, fmt.Sprintf("Table %s not found.", in.TableID))
		}
		if err := auth.AssertProjectAccess(ctx, userID, projectID); err != nil {
			return fail(err)
		}

		whiteboardID, err := getWhiteboardIDForTable(ctx, in.TableID)
		if err != nil {
			return fail(err)
		}

		ack, err := socket.SocketEmitWithAck(ctx, whiteboardID, userID, "column:reorder", map[string]any{
			"tableId":          in.TableID,
			"orderedColumnIds": in.OrderedColumnIDs,
		})
		if err != nil {
			return fail(err)
		}
		if !ack.OK() {
			return mcpError(mcperr.AckCodeToMcpCode(ack.Code()), msgOr(ack.Message(), "Server rejected column reorder."))
		}
		if e := ack.Entity(); e != nil {
			return success(e)
		}
		return success(map[string]any{"tableId": in.TableID, "orderedColumnIds": in.OrderedColumnIDs})
	})
}
